File: server/src/Player.js
function Player(socket, name) {
    this.socket = socket;
    this.id = socket.id;
    this.name = name;
    this.emit = this.socket.emit.bind(socket);
    this.on = this.socket.on.bind(socket);
    this.connected = true;
}

Player.prototype.disconnect = function () {
    if (this.connected) {
        this.socket.disconnect();
        this.connected = false;
        console.log(this.getDescription() + ' disconnected');
    }
};

Player.prototype.getDescription = function () {
    return this.name + '(#' + this.id + ')';
};

module.exports = Player;

File: server/src/main/kotlin/cn/newinfinideas/myomyw/Player.kt
package cn.newinfinideas.myomyw

import com.google.gson.GsonBuilder

class Player(private val socket: Bus, val name: String, private val id: Int) {
    val builder = GsonBuilder()

    inline fun <reified T> onPacket(crossinline func: suspend (T) -> Unit) {
        this.on(PacketTable.getName(T::class)) { func(builder.create().fromJson(it, T::class.java)) }
    }

    fun on(event: String, func: suspend (String) -> Unit) {
        this.socket.on(event, true, func)
    }

    suspend inline fun <reified T> emit(content: T) = emit(PacketTable.getName(T::class), builder.create().toJson(content))

    suspend fun emit(name: String, content: String) = this.socket.emit(name, content)

    suspend fun disconnect() = this.socket.disconnect()

    fun getDescription(): String = "${this.name}(${this.id})"
}